purchase_order_supplierinfo_update/models/purchase_order.py

```diff
@@ -0,0 +1,72 @@
+# Copyright 2021 Tecnativa - Ernesto Tejeda
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+
+from odoo import api, fields, models
+
+
+class PurchaseOrder(models.Model):
+    _inherit = "purchase.order"
+
+    def write(self, vals):
+        no_updated = self.filtered(lambda r: r.state not in ["purchase", "done"])
+        res = super().write(vals)
+        if vals.get("state", "") in ["purchase", "done"]:
+            no_updated.mapped("order_line").update_supplierinfo_price()
+        return res
+
+
+class PurchaseOrderLine(models.Model):
+    _inherit = "purchase.order.line"
+
+    @api.model_create_multi
+    def create(self, vals_list):
+        res = super().create(vals_list)
+        res.update_supplierinfo_price()
+        return res
+
+    def write(self, vals):
+        res = super().write(vals)
+        if vals.get("price_unit"):
+            self.update_supplierinfo_price()
+        return res
+
+    def update_supplierinfo_price(self):
+        for line in self.filtered(
+            lambda r: not r.display_type and r.order_id.state in ["purchase", "done"]
+        ):
+            domain = [
+                ("partner_id", "=", line.partner_id.id),
+                ("product_id", "=", line.product_id.id),
+                ("date_order", ">", line.date_order),
+            ]
+            if not self.env["purchase.order.line"].search(domain, limit=1):
```

Comment:
I think this can be done with `search_count` instead of `search`. It's more efficient if you don't need any fields from the search result.
```suggestion
            if not self.env["purchase.order.line"].search_count(domain):
```